javascript/ql/src/semmle/javascript/NullSensitiveContext.qll

```diff
@@ -0,0 +1,488 @@
+/**
+ * Provides classes for working with contexts sensitive to containing null
+ * and undefined.
+ */
+
+import javascript
+
+/**
+ * A `NullSensitiveContext` is a syntactic context in which neither null nor
+ * undefined should occur, because the context is guaranteed to inspect its
+ * content and thereby throw an error or produce other undesirable behavior.
+ *
+ * The expression found in that context is called the principal argument of
+ * the context.
+ *
+ * In examples, we denote the position of the principal argument by use of
+ * an underscore. For example, in `_ + 0`, the context's principle argument
+ * is the left operand of the addition.
+ */
+abstract class NullSensitiveContext extends Expr {
+  abstract Expr getPrincipalArgument();
```

Comment:
Needs qldoc.

Reply:
> On second thought, I'd suggest replacing all of these abstract classes with a simple predicate on `Expr`. It's just as extensible (since you can override that predicate when needed), and if you combine the individual cases as suggested in the review comments it won't even be that complicated.

Would this make it possible to include the `for-of` statement iteration domain? Or should we omit that in preference of a method predicate?

Reply:
The method specifically?

Reply:
Yes, we generally provide qldoc for all public predicates (and ideally for private ones as well).